hedera-node/hedera-app/src/main/java/com/hedera/node/app/services/ServiceProviderImpl.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright (C) 2023 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.node.app.services;
+
+import com.hedera.node.app.spi.FacilityFacade;
+import com.hedera.node.app.spi.service.Service;
+import com.hedera.node.app.spi.service.ServiceFactory;
+import com.hedera.node.app.spi.service.ServiceProvider;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.ServiceLoader;
+import java.util.ServiceLoader.Provider;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class ServiceProviderImpl implements ServiceProvider {
+
+    private final Map<Class<Service>, Service> services;
+
+    public ServiceProviderImpl(@NonNull final FacilityFacade facilityFacade) {
+        Objects.requireNonNull(facilityFacade, "FacilityFacade can not be null");
+        services = new HashMap<>();
+        final ServiceLoader<ServiceFactory> serviceLoader = ServiceLoader.load(ServiceFactory.class);
+        final Set<ServiceFactory> servicesFactories =
+                serviceLoader.stream().map(Provider::get).collect(Collectors.toSet());
+
+        servicesFactories.stream()
+                .sorted((a, b) -> sort(a, b))
+                .forEach(factory -> createAndAddService(facilityFacade, factory));
+    }
+
+    private void createAndAddService(final FacilityFacade facilityFacade, final ServiceFactory factory) {
+        if (!services.keySet().containsAll(factory.getDependencies())) {
+            throw new IllegalStateException("Can not add service for " + factory.getServiceClass()
+                    + " since the service has a missing dependency");
+        }
+        services.keySet().stream()
+                .filter(service -> dependOnEachOther(service, factory.getServiceClass()))
+                .anyMatch(serviceClass -> {
+                    throw new IllegalStateException("Can not add service for " + factory.getServiceClass()
+                            + " since the service has a clash with " + serviceClass);
+                });
+        final Service service = factory.createService(this, facilityFacade);
```

Comment:
I think we can check duplicate services and circular dependencies by comparing `serviceClass` and `factory.getServiceClass()` using a for-loop like below. Creating two streams will sacrifice some performance than using a for-loop in this case. Also a service can be created at the last step after validation.

```
services.forEach((serviceClass, service) -> {
	// check duplicate services

	// check circular dependencies
});

// create a service
```

Reply:
No idea how I should do the dependency check without containsAll

Reply:
I meant these two streams, `services.keySet().stream()` (L54) and `services.values().stream()` (L62).